Hack to fix conda build failure caused by incorrect libthrift dep

arthurp opened the pull request:
This may not need to be merged if https://github.com/conda-forge/arrow-cpp-feedstock/issues/461 is fixed before we have problems.

ddn0 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -1645,88 +1783,93 @@ BuildFinalEdges(GraphState* builder, WriterProperties* properties) {
+  std::cout << "Edges: " << topology->out_dests->length() << std::endl;
+  std::cout << "Edge Properties: " << final_edge_table->num_columns()
+            << std::endl;
+  std::cout << "Edge Types: " << final_type_table->num_columns() << std::endl;
```
std::endl vs "\n"

patrickkenney9801 commented on tools/graph-convert/tests/CMakeLists.txt:
```diff
@@ -6,6 +6,14 @@ set_tests_properties(unit-time-parser PROPERTIES LABELS quick)
 target_link_libraries(graph-properties-convert-test PRIVATE LLVMSupport)
 target_link_libraries(graph-properties-convert-test PRIVATE graph-properties-convert-common)
+<<<<<<< HEAD
+=======
```
there are not any tests for functionality currently, I can just do some toy examples for parsing some in memory documents like you suggested but they will be unable to test actual mongodb interactions unless I find a remote read-only database or we set one up ourselves

ddn0 replied:
I think having some tests for the conversion process will be useful even without the mongodb interactions. It breaks down the correctness assumption into two manageable and independent pieces: (1) can we create documents from a connection and (2) given some documents can we convert them.

Depending on your estimation on value of getting coverage for (1), you could make a short test which when given the address of a mongodb instance checks that documents come out in a certain form. Not having used mongodb, I'd assume that these types of tests aren't as valuable compared to (2) because all you'd be testing is how closely mongodb fulfills its own API (or our understanding of it).

patrickkenney9801 replied:
True, will just do 2 then

ddn0 commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -1304,6 +1143,164 @@ GetUserInput(mongoc_database_t* database,
+    // handle document
+    while (bson_iter_next(&iter)) {
+      const bson_value_t* elt = bson_iter_value(&iter);
+      auto name = std::string(bson_iter_key(&iter), bson_iter_key_len(&iter));
```
This is a funny way to write:
```
std::string name{bson_iter_key(....), .....};
```

ddn0 commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -1304,6 +1143,164 @@ GetUserInput(mongoc_database_t* database,
+    // handle document
+    while (bson_iter_next(&iter)) {
+      const bson_value_t* elt = bson_iter_value(&iter);
+      auto name = std::string(bson_iter_key(&iter), bson_iter_key_len(&iter));
```
ditto

ddn0 commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -0,0 +1,1372 @@
+        }
+
+        // since all edge cases have been checked, we can add this property
+        auto keyIter = builder->node_properties.keys.find(elt_name);
```
key_iter

ddn0 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -1871,62 +1529,60 @@ galois::GraphComponents galois::ConvertGraphML(const std::string& infilename,
   auto topology = std::make_shared<galois::graphs::GraphTopology>();
   arrow::Status st;
-  std::shared_ptr<arrow::UInt64Builder> topology_indices_builder =
+  std::shared_ptr<arrow::UInt64Builder> topologyIndicesBuilder =
```
topology_indices_builder

ddn0 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -1871,62 +1529,60 @@ galois::GraphComponents galois::ConvertGraphML(const std::string& infilename,
 
-/// writePropertyGraph formally builds katana form via PropertyFileGraph from
-/// imported components and writes the result to target directory
+/// convertToPropertyGraphAndWrite formally builds katana form via
```
/// WritePropertyGraph formally builds ....?

tylershunt commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -0,0 +1,1371 @@
+};
+
+struct bson_value_t_wrapper {
+  bson_value_t val;
```
Maybe for the future since we've already got this code here but `nlohmman/json` [can parse `bson` as an input format](https://github.com/nlohmann/json#binary-formats-bson-cbor-messagepack-and-ubjson). This is assuming that Mongo's `bson` is the same `bson` that the rest of the world uses. Not exactly sure if the mapping makes sense but if you need to load serialized `bson` objects into custom data structures I've found `nlohmann` to be pretty good!

tylershunt commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -0,0 +1,1371 @@
+  std::set<std::string> embedded_relations;
+};
+
+struct bson_value_t_wrapper {
```
Apologies if someone already asked but what is this wrapper for?

patrickkenney9801 replied:
C++ just does not like having a bson_value_t in a vector so it needs a wrapper

tylershunt commented on tools/graph-convert/graph-properties-convert-mongodb.cpp:
```diff
@@ -0,0 +1,1371 @@
+/**********************************/
+
+// mongoc_init() should be called before this function
+mongoc_client_t* GetMongoClient(const char* uri_string) {
```
Feel free to ignore this but I just thought that I'd mention:
A pattern we're using in other places to manage resources that are allocated/freed in the c style is to wrap them in an object that calls their free function when it is destroyed; something like [RDGFile](https://github.com/KatanaGraph/katana/blob/a1bf3f9c83c6853a1ea37082eca88b5b9ce224a9/libtsuba/include/tsuba/RDG.h#L24)

patrickkenney9801 replied:
Good point, I will fix this in the near future and look for other similar cases

nicelhc13 commented on tools/graph-convert/CMakeLists.txt:
```diff
@@ -60,14 +60,34 @@ install(TARGETS graph-convert-huge
 target_link_libraries(graph-properties-convert-common PRIVATE LibXml2::LibXml2)
 target_link_libraries(graph-properties-convert-common PUBLIC galois_shmem date::date)
 
+if(mongoc-1.0_FOUND)
```
Is there any specific reason to fix the version to 1.0?

patrickkenney9801 replied:
that's just what it is called, I actually retrieve version 1.6

nicelhc13 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -139,6 +85,90 @@ std::shared_ptr<arrow::Table> BuildTable(std::vector<ArrowArrays>* chunks,
+  auto builder = std::make_shared<T>();
+  auto st      = builder->AppendNulls(elts);
+  null_map->insert(std::pair<int, std::shared_ptr<arrow::Array>>(
+      builder->type()->id(), BuildArray(builder)));
```
Just small opinion: define `int id = builder->type()->id()` and replacing two deference looks better for me if you don't have any reason.

nicelhc13 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -139,6 +85,90 @@ std::shared_ptr<arrow::Table> BuildTable(std::vector<ArrowArrays>* chunks,
+  // extraneous info
+  auto builder = std::make_shared<arrow::TimestampBuilder>(type, pool);
+  auto st      = builder->AppendNulls(elts);
+  null_map->insert(std::pair<int, std::shared_ptr<arrow::Array>>(
```
Same comment as above.

nicelhc13 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -887,6 +481,52 @@ uint64_t SetEdgeId(TopologyState* topology_builder,
+void ResolveIntermediateIDs(GraphState* builder) {
+  TopologyState* topology = &builder->topology_builder;
+
+  for (size_t i = 0; i < topology->destinations.size(); i++) {
```
Minor comment: [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement) recommends to use prefix form (`++i`)

nicelhc13 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -887,6 +481,52 @@ uint64_t SetEdgeId(TopologyState* topology_builder,
+        dest = builder->nodes;
+        topology->node_indexes.insert(
+            std::pair<std::string, size_t>(str_id, dest));
+        builder->nodes++;
```
Same comment as above.

nicelhc13 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -1403,419 +1108,78 @@ KeyGraphML ProcessKey(xmlTextReaderPtr reader) {
+        for_node = xmlStrEqual(value, BAD_CAST "node") == 1;
+        for_edge = xmlStrEqual(value, BAD_CAST "edge") == 1;
+      } else if (xmlStrEqual(name, BAD_CAST "attr.label")) {
+        attr_label = std::string((const char*)value);
```
How about using `static_cast`?

nicelhc13 commented on tools/graph-convert/graph-properties-convert.cpp:
```diff
@@ -1825,23 +1189,35 @@ galois::GraphComponents galois::ConvertGraphML(const std::string& infilename,
-          ProcessGraph(reader, &builder, &properties);
-          finishedGraph = true;
+          std::cout << "Finished processing headers\n";
+          finished_header = true;
```
Is there any reason to use `finished_header` instead of `break`? I also saw similar code on GraphML converter.

patrickkenney9801 replied:
the xml reader I am using does not allow "jumping around" so even if I do not care about a certain xml element or other, I still have to read it

nicelhc13 commented on tools/graph-convert/tests/CMakeLists.txt:
```diff
@@ -5,19 +5,29 @@ set_tests_properties(unit-time-parser PROPERTIES LABELS quick)
+
 add_test(NAME convert-properties-graphml
-  COMMAND graph-properties-convert-test --graphml --movies ${CMAKE_CURRENT_SOURCE_DIR}/../test-inputs/movies.graphml
+  COMMAND graph-properties-convert-test --neo4j --movies ${CMAKE_CURRENT_SOURCE_DIR}/../test-inputs/movies.graphml
```
In the past, we need to specify both `neo4j` and `graphml` flags, but now don't we need `graphml`?

arthurp committed:
Hack to fix conda build failure caused by incorrect libthrift dep

Revert after https://github.com/conda-forge/arrow-cpp-feedstock/issues/461
is closed.

arthurp commented:
Merging despite master building correctly. It hit me that there may well be some RNG involved in the failure because of how SAT solving and conda works. Better safe than sorry.
